File: src/med_exam_toolkit/ai/result.py
"""AI 结果解析与验证（答案/解析补全）"""
from __future__ import annotations

import json
import logging
import re
from typing import Any

from med_exam_toolkit.models import SubQuestion

logger = logging.getLogger(__name__)


def parse_response(raw: str) -> dict[str, Any]:
    text = (raw or "").strip()
    if not text:
        return {}

    # 策略 1：直接 JSON
    try:
        return json.loads(text)
    except json.JSONDecodeError:
        pass

    # 策略 2：markdown 代码块
    m = re.search(r"```(?:json)?\s*(\{.*?\})\s*```", text, re.DOTALL)
    if m:
        try:
            return json.loads(m.group(1))
        except json.JSONDecodeError:
            pass

    # 策略 3：提取首个 JSON 对象
    m = re.search(r"\{.*\}", text, re.DOTALL)
    if m:
        try:
            return json.loads(m.group(0))
        except json.JSONDecodeError:
            pass

    logger.warning("AI 返回内容无法解析为 JSON: %s", text[:200])
    return {}


def validate_result(
    result: dict[str, Any],
    *,
    need_answer: bool,
    need_discuss: bool,
) -> tuple[bool, list[str]]:
    missing: list[str] = []
    if need_answer and not str(result.get("answer", "")).strip():
        missing.append("answer")
    if need_discuss and not str(result.get("discuss", "")).strip():
        missing.append("discuss")
    return (len(missing) == 0), missing


def apply_to_subquestion(
    sq: SubQuestion,
    result: dict[str, Any],
    *,
    model_name: str,
    overwrite: bool = False,
) -> None:
    if not result:
        return

    ai_answer = str(result.get("answer", "")).strip()
    ai_discuss = str(result.get("discuss", "")).strip()

    try:
        ai_confidence = float(result.get("confidence", 0.0) or 0.0)
    except (TypeError, ValueError):
        ai_confidence = 0.0

    if ai_answer:
        sq.ai_answer = ai_answer
    if ai_discuss:
        sq.ai_discuss = ai_discuss
    sq.ai_confidence = ai_confidence
    sq.ai_model = model_name
    sq.ai_status = "pending"

    if overwrite:
        if ai_answer and not (sq.answer or "").strip():
            sq.answer = ai_answer
        if ai_discuss and not (sq.discuss or "").strip():
            sq.discuss = ai_discuss

File: internal/ai/result.go
package ai

import (
	"encoding/json"
	"regexp"
	"strconv"
	"strings"

	"github.com/zyh001/med-exam-kit/internal/models"
)

// ParseResponse extracts a JSON object from AI output, handling markdown fences etc.
func ParseResponse(raw string) map[string]any {
	text := strings.TrimSpace(raw)
	if text == "" {
		return nil
	}

	// Strategy 1: direct JSON
	var m map[string]any
	if err := json.Unmarshal([]byte(text), &m); err == nil {
		return m
	}

	// Strategy 2: markdown code block
	reFence := regexp.MustCompile("(?s)```(?:json)?\\s*(\\{.*?\\})\\s*```")
	if matches := reFence.FindStringSubmatch(text); len(matches) > 1 {
		if err := json.Unmarshal([]byte(matches[1]), &m); err == nil {
			return m
		}
	}

	// Strategy 3: first JSON object
	reObj := regexp.MustCompile(`(?s)\{.*\}`)
	if match := reObj.FindString(text); match != "" {
		if err := json.Unmarshal([]byte(match), &m); err == nil {
			return m
		}
	}

	return nil
}

// ValidateResult checks whether the result has the required fields.
// Returns (ok, missingFields).
func ValidateResult(result map[string]any, needAnswer, needDiscuss bool) (bool, []string) {
	var missing []string
	if needAnswer {
		a, _ := result["answer"].(string)
		if strings.TrimSpace(a) == "" {
			missing = append(missing, "answer")
		}
	}
	if needDiscuss {
		d, _ := result["discuss"].(string)
		if strings.TrimSpace(d) == "" {
			missing = append(missing, "discuss")
		}
	}
	return len(missing) == 0, missing
}

// ApplyToSubQuestion writes AI result into the SubQuestion fields.
// If overwrite is true, also fills empty official answer/discuss from AI.
func ApplyToSubQuestion(sq *models.SubQuestion, result map[string]any, modelName string, overwrite bool) {
	if result == nil {
		return
	}

	aiAnswer := strings.TrimSpace(strAny(result["answer"]))
	aiDiscuss := strings.TrimSpace(strAny(result["discuss"]))

	var aiConfidence float64
	switch v := result["confidence"].(type) {
	case float64:
		aiConfidence = v
	case string:
		aiConfidence, _ = strconv.ParseFloat(v, 64)
	}

	if aiAnswer != "" {
		sq.AIAnswer = aiAnswer
	}
	if aiDiscuss != "" {
		sq.AIDiscuss = aiDiscuss
	}
	sq.AIConfidence = aiConfidence
	sq.AIModel = modelName
	sq.AIStatus = "pending"

	if overwrite {
		if aiAnswer != "" && strings.TrimSpace(sq.Answer) == "" {
			sq.Answer = aiAnswer
		}
		if aiDiscuss != "" && strings.TrimSpace(sq.Discuss) == "" {
			sq.Discuss = aiDiscuss
		}
	}
}

func strAny(v any) string {
	if s, ok := v.(string); ok {
		return s
	}
	return ""
}
